src/fciqmc.jl

```diff
@@ -446,23 +456,52 @@ Returns the result `ṽ`, a (possibly changed) reference to working memory `w̃`
 `stats = [spawns, deaths, clones, antiparticles, annihilations]`. Stats will
 contain zeros when running in deterministic mode.
 """
-function fciqmc_step!(Ĥ, v::D, shift, dτ, pnorm, w::D;
-                      m_strat::MemoryStrategy = NoMemory()) where D
-    # serial version
+function fciqmc_step!(Ĥ, dv, shift, dτ, pnorm, w, m = 1.0;
+                      m_strat::MemoryStrategy = NoMemory())
+    # single-threaded version suitable for MPI
+    # m is an optional dummy argument that can be removed later
+    v = localpart(dv)
     @assert w ≢ v "`w` and `v` must not be the same object"
-    stats = zeros(Int, 5) # pre-allocate array for stats
-    zero!(w) # clear working memory
-    for (add, num) in pairs(v)
-        res = fciqmc_col!(w, Ĥ, add, num, shift, dτ)
-        stats .+= res # just add all stats together
-    end
+    empty!(w) # clear working memory
+    # call fciqmc_col!() on every entry of `v` and add the stats returned by
+    # this function:
+    stats = sum(p-> SVector(fciqmc_col!(w, Ĥ, p.first, p.second, shift, dτ)),
+      pairs(v))
     r = applyMemoryNoise!(w, v, shift, dτ, pnorm, m_strat) # memory noise
-    # norm_project!(w, p_strat) # project coefficients of `w` to threshold
-    # thresholdProject!(w, v, shift, dτ, m_strat) # apply walker threshold if applicable
-    return w, v, stats, r
+    return (sort_into_targets!(dv, w, stats)... , r) # MPI syncronizing
     # stats == [spawns, deaths, clones, antiparticles, annihilations]
 end # fciqmc_step!
 
+# function fciqmc_step!(Ĥ, v::D, shift, dτ, pnorm, w::D;
+#                       m_strat::MemoryStrategy = NoMemory()) where D
+#     # serial version
+#     @assert w ≢ v "`w` and `v` must not be the same object"
+#     zero!(w) # clear working memory
+#     # call fciqmc_col!() on every entry of `v` and add the stats returned by
+#     # this function:
+#     stats = mapreduce(p-> SVector(fciqmc_col!(w, Ĥ, p.first, p.second, shift, dτ)), +,
+#       pairs(v))
+#     r = applyMemoryNoise!(w, v, shift, dτ, pnorm, m_strat) # memory noise
+#     return w, v, stats, r
+#     # stats == [spawns, deaths, clones, antiparticles, annihilations]
+# end # fciqmc_step!
+
+# provide allocation of statss array for multithreading as a separate function in order to
```

Comment:
What is a "statss" array?

Reply:
`statss` is a Vector with `nt` slots, each collecting the `stats` returned by `fciqmc_col`, where `nt` is the number of threads. The vector is needed such that each thread can write into its own slot and race conditions can be avoided.